### ISSUE TITLE
Suggestion: Decouple the registry from auth and the client?

### ISSUE THREAD

**Issue by jwietelmann** [coryodaniel/k8s#44]
The thing stopping me from using `k8s` right now is the registry. My clusters have [dex](https://github.com/dexidp/dex) for user auth, and my app gets auth tokens from there. My users do not share Kubernetes credentials when interacting with clusters; they each get their own session tokens via OAuth and use those.

If the relationship between configuration and the registry and auth were decoupled a bit, such that I didn't have to necessarily use it, I would be able to benefit from using `K8s.Client`. Such as it currently is, it seems like I would have to register a cluster per set of credentials? (Please correct me if that's not right.)

This is ideally how I'd like to be able to interact with `k8s`:

```elixir
%{
  api_host: "https://example.com:6443",
  # And more, but not auth credentials
}
|> Client.new(token: "my user's token")
|> Client.list("apps/v1", :deployment, namespace: "default")
|> Client.run()
```

I don't like this quite as much, but it might be a shorter leap from how `k8s` is now, and it would solve my problem I think:

```elixir
prod = K8s.Conn.new(some_config_without_credentials)
K8s.Cluster.Registry.add(:prod, prod)

operation = Client.list("apps/v1", :deployment, namespace: "default")
{:ok, deployments} = Client.run(operation, :prod, token: "my user's token")
```

**Comment by jwietelmann** [comment]
Oh, and I do see that there is this `AuthProvider` thing, and I'm still trying to understand the code, but to be clear: I get my users' tokens via OAuth and store them in the session (this is a Phoenix app), and I'm not sure how the `AuthProvider` would ever be able to get the token out of a `Plug.Conn` that's being constructed at runtime for every HTTP request.

**Comment by coryodaniel** [comment]
I think OAuth would be difficult via the auth provider...

As far as decoupling, Ive been thinking about this recently. I was considering adding a second function that takes an auth config and then have run/5 lookup configs by name and then pass them to the function that explicitly takes a config.

Related:
#42 
#43

**Comment by coryodaniel** [comment]
Beginning JIT refactoring this weekend. Recapping on this just to make sure I incorporate it.

_Support for overriding the auth for a cluster at request time._

1. Cluster's auth would be optional, currently no config, means use pod service account. I think that is fine to leave as a default, such that its not a breaking change
2. Accept an additional `opt` on `run/N` that overrides the auth that was registered for the cluster
  a. reconsider `opt` structure as it is becoming a bit of a _pile on_
  b. with JIT, the provided credentials would be used for discovery

**Comment by coryodaniel** [comment]
@jwietelmann I'm working on the JIT stuff and I plan to incorporate ad-hoc `conn`ections instead of always using a cluster name atom.

I figured Id share some details on implementing an a custom auth strategy.

A few notes first:
1. [`AuthProvider`](https://github.com/coryodaniel/k8s/blob/master/lib/k8s/conn/auth/auth_provider.ex) is itself an authentication strategy that allow shell calls to provide a Bearer Token. It's unfortunately named, but the names of the modules follow the key names in a kube config file. More on this strategy can be found [here](https://banzaicloud.com/blog/kubeconfig-security/).
2. The [`K8s.Conn`](https://github.com/coryodaniel/k8s/blob/master/lib/k8s/conn.ex#L58) struct encapsulates a connection to a cluster. It has the cluster address as well as how to authenticate to the cluster. These `Conn`s can be constructed manually, but there are a few helpers here to create one.
3.   [Token auth strategy](https://github.com/coryodaniel/k8s/blob/master/lib/k8s/conn/auth/token.ex#L13-L14) is probably the simplest strategy.

Two things are required to implement a custom auth strategy:

1. Implement this [behaviour](https://github.com/coryodaniel/k8s/blob/master/lib/k8s/conn/auth.ex) for auth strategies. It should return `nil` if the strategy doesnt apply to the config provided, otherwise, return struct
2. Implement this [protocol](https://github.com/coryodaniel/k8s/blob/master/lib/k8s/conn/request_options.ex#L19) which should create a `RequestOptions` struct with HTTP Headers and SSL connection details.

Given the [Token example](https://github.com/coryodaniel/k8s/blob/master/lib/k8s/conn/auth/token.ex#L13-L14)
* Line 13 implements the case where this auth strategy would be able to generate request options
* Line 14 implements the default case where it cannot authenticate the request
* Lines 19-24 implement how to generate HTTP Headers and SSL options to be used by HTTPoison to make the HTTP requests.

Once this refactor is complete you'll be able to do:

```elixir
# you'll need to build this conn struct yourself
conn = %K8s.Conn{...} 
op = K8s.Client.get(...)
{:ok, response} = K8s.Client.run(op, conn)
```

I plan to keep support for supplying the cluster names as well:

```elixir
operation = K8s.Client.get(...)
{:ok, response} = K8s.Client.run(operation, :dev)
```

**Comment by coryodaniel** [comment]
This was released in 0.5 @jwietelmann
